envoy/datadog_checks/envoy/check.py

```diff
@@ -0,0 +1,442 @@
+        'new_name': 'watchdog_miss.count',
+    },
+    r'^envoy_(.+\_.+)_watchdog_mega_miss$': {
+        'label_name': 'thread_name',
+        'metric_type': 'monotonic_count',
+        'new_name': 'watchdog_mega_miss.count',
+    },
+    r'^envoy_cluster_circuit_breakers_(\w+)_cx_open$': {
+        'label_name': 'priority',
+        'metric_type': 'gauge',
+        'new_name': 'cluster.circuit_breakers.cx_open',
+    },
+    r'^envoy_cluster_circuit_breakers_(\w+)_cx_pool_open$': {
+        'label_name': 'priority',
+        'metric_type': 'gauge',
+        'new_name': 'cluster.circuit_breakers.cx_pool_open',
+    },
+    r'^envoy_cluster_circuit_breakers_(\w+)_rq_open$': {
+        'label_name': 'priority',
+        'metric_type': 'gauge',
+        'new_name': 'cluster.circuit_breakers.rq_open',
+    },
+    r'^envoy_cluster_circuit_breakers_(\w+)_rq_pending_open$': {
+        'label_name': 'priority',
+        'metric_type': 'gauge',
+        'new_name': 'cluster.circuit_breakers.rq_pending_open',
+    },
+    r'^envoy_cluster_circuit_breakers_(\w+)_rq_retry_open$': {
+        'label_name': 'priority',
+        'metric_type': 'gauge',
+        'new_name': 'cluster.circuit_breakers.rq_retry_open',
+    },
+    r'^envoy_listener_admin_(.+\_.+)_downstream_cx_active$': {
+        'label_name': 'handler',
+        'metric_type': 'gauge',
+        'new_name': 'listener.admin.downstream_cx_active',
+    },
+    r'^envoy_listener_(.+\_.+)_downstream_cx_active$': {
+        'label_name': 'handler',
+        'metric_type': 'gauge',
+        'new_name': 'listener.downstream_cx_active',
+    },
+    r'^envoy_listener_admin_(.+\_.+)_downstream_cx$': {
+        'label_name': 'handler',
+        'metric_type': 'monotonic_count',
+        'new_name': 'listener.admin.downstream_cx.count',
+    },
+    r'^envoy_listener_(.+)_downstream_cx$': {
+        'label_name': 'handler',
+        'metric_type': 'monotonic_count',
+        'new_name': 'listener.downstream_cx.count',
+    },
+}
+
+
+class EnvoyCheck(OpenMetricsBaseCheckV2):
+    __NAMESPACE__ = 'envoy'
+
+    DEFAULT_METRIC_LIMIT = 0
+
+    def __init__(self, name, init_config, instances):
+        super().__init__(name, init_config, instances)
+        self.check_initializations.append(self.configure_additional_transformers)
+
+    def get_default_config(self):
+        return {
+            'metrics': [METRIC_MAP],
+            'rename_labels': LABEL_MAP,
+        }
+
+    def configure_transformer_label_in_name(self, metric_pattern, new_name, label_name, metric_type):
+        method = getattr(self, metric_type)
+
+        def transform(metric, sample_data, runtime_data):
+            for sample, tags, hostname in sample_data:
+                parsed_sample_name = sample.name
+                if sample.name.endswith("_total"):
+                    parsed_sample_name = re.match("(.*)_total$", sample.name).groups()[0]
+
+                compiled_name = re.compile(metric_pattern)
```

Comment:
Let's cache the compiled patterns in the outer scope using `collections.defaultdict`